app/models/concerns/labware_creators/donor_pooling_calculator.rb

```diff
@@ -130,34 +130,110 @@ def unique_donor_ids(group)
     group.map { |well| well.aliquots.first.sample.sample_metadata.donor_id }.uniq
   end
 
-  # Distributes samples across pools based on group sizes. It sorts the groups
-  # by size and splits the largest group into two until the number of groups
-  # equals the number of pools or until all groups have a size of 1. The input
-  # groups are the result of applying conditions, hence they cannot be mixed.
+  # Validates the number of pools requested by the user. The number of pools must
+  # be between 1 and 8, inclusive. The total number of wells must be divisible by
+  # the number of pools, with a difference of at most 1 well.
   #
-  # If the request number of pools is 6 and the input groups are
-  # [[1, 2, 3], [4, 5], [6, 7, 8, 9]] where the numbers denote wells,
+  def validate_number_of_pools(wells, number_of_pools)
+    total_wells = wells.size
+
+    return unless total_wells < number_of_pools || total_wells > number_of_pools * ((total_wells / number_of_pools) + 1)
+    raise "Cannot distribute #{total_wells} wells into #{number_of_pools} pools such that the difference is at most 1."
+  end
+
+  def handle_non_unique_donor_id(depth, number_of_pools, donor_id)
+    return unless depth == number_of_pools
+    raise "Unable to allocate well with donor ID #{donor_id}. All pools contain this donor."
+  end
+
+  # Recursive function to assign wells to pools
+  def assign_well_to_pool(args)
+    well, pools, used_donor_ids, pool_index, number_of_pools, depth =
+      args.values_at(:well, :pools, :used_donor_ids, :pool_index, :number_of_pools, :depth)
+
+    donor_id = well.aliquots.first.sample.sample_metadata.donor_id
+
+    if donor_already_used?(donor_id, used_donor_ids, pool_index)
+      handle_conflict_donor_ids(donor_id, args, depth, number_of_pools, pool_index)
+    else
+      add_to_pool(donor_id, used_donor_ids, pool_index, pools, well)
+    end
+  end
+
+  def donor_already_used?(donor_id, used_donor_ids, pool_index)
+    used_donor_ids[pool_index].include?(donor_id)
+  end
+
+  def handle_conflict_donor_ids(donor_id, args, depth, number_of_pools, pool_index)
```

Comment:
I'd prefer to see method comments throughout.
If using VS code with codepilot chat it does a pretty good job if you highlight a method and say 'add method comment'

Reply:
you have good clear method names mind.

Reply:
and the complex methods are commented